### PR TITLE
BrokerHeartbeatManager: Schedule heartbeats

### PR DESCRIPTION
**Pull request by aloknnikhil**
- [x] First pass for Broker Heartbeats
- [x] Implements a broker heartbeat scheduler that communicates requested state changes to the active controller
- [x] Implements a future/promise when the state change is complete/fails
- [x] Only wired to send state changes
- [x] Basic unit testing for state changes in-flight (single/multiple/parallel/error)
- [x] Integrate w/ Broker registration
- [x] Integrate w/ BrokerMetadataListener
- [x] Controller-side API
- [ ] Wiring up/integrating w/ KafkaServer to drive the state changes
- [ ] Add more timeouts configuration 
    - [x] Broker side: 
        - [x] Time between heartbeats ➝ `registration.heartbeat.interval.ms`
        - [x] Time after which the broker fences itself ➝ `registration.lease.timeout.ms`
    - [ ] Controller Side:
        - [ ] Time to boot a broker from the cluster ➝ `registration.lease.timeout.ms`
        - [ ] Time after which the broker ID is up for grabs 
                ➝ `registration.heartbeat.interval.ms` < x < `registration.lease.timeout.ms`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cmccabe** [comment]
Thanks for working on this, @aloknnikhil !

I think it would be better to have a dedicated thread for this, rather than re-using an existing scheduler.  The issue is that we really don't want anything else to block the heartbeats.

I think we also want to measure time using time.nanoseconds rather than time.milliseconds.   The Time#milliseconds method can be affected by clock adjustments.

One more thing is that we recently split the broker heartbeat into a registration message, and a heartbeat.  So we probably want to manage the registration here as well.  So maybe this should have a more general name like BrokerLifecycleManager ?

**Comment by aloknnikhil** [comment]
> Thanks for working on this, @aloknnikhil !
> 
> I think it would be better to have a dedicated thread for this, rather than re-using an existing scheduler. The issue is that we really don't want anything else to block the heartbeats.
> 
> I think we also want to measure time using time.nanoseconds rather than time.milliseconds. The Time#milliseconds method can be affected by clock adjustments.
> 
> One more thing is that we recently split the broker heartbeat into a registration message, and a heartbeat. So we probably want to manage the registration here as well. So maybe this should have a more general name like BrokerLifecycleManager ?

- Dedicated scheduler for the heartbeats: Makes sense. I'll leave the current constructor as it makes it easy to unit test w/ the mock scheduler. I'll pass in a dedicated scheduler at the time of construction when ever we integrate this w/ KafkaServer

- Time.nano v/s Time.milli: Yea, I agree too considering all time references in the code are w/ respective to the same JVM. I had a TODO to switch to nano, but seems to have disappeared in some refactor.

- BrokerLifecycleManager: Fair enough. I assumed BrokerRegistration was handled by a separate entity. But makes sense to keep it all in one place which fully owns the BrokerState management.

**Comment by aloknnikhil** [comment]
Addressed the changes. I am going to continue working on integrating the Broker Registration into the lifecycle and the controller-side APIs @cmccabe

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+      validateResponse(response) match {
+        case None =>
+          // Check for API errors
+          val body = response.responseBody().asInstanceOf[BrokerRegistrationResponse]
+          Errors.forCode(body.data().errorCode()) match {
+            case Errors.DUPLICATE_BROKER_REGISTRATION =>
+              currentState = metadata.BrokerState.NOT_RUNNING
+              promise.tryFailure(Errors.DUPLICATE_BROKER_REGISTRATION.exception())
+            case Errors.NONE =>
+              // Registration success
```
Comment (rondagostino) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:160]: The heartbeat needs to send a successful registration event to the broker's metadata listener via something like this:

```
kafkaBroker.brokerMetadataListener.put(OutOfBandRegisterLocalBrokerEvent(brokerEpoch))
```

The broker metadata listener code was just merged.

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+      }
+
+      // Check for any pending state changes that have been queued
+      var state = requestQueue.poll
+      if (state == null) {
+        // No state change requests queued
+        if (timeSinceLastHeartbeat > config.registrationLeaseTimeoutMs) {
+          error(s"Last successful heartbeat was $timeSinceLastHeartbeat ms ago")
+          // Fence ourselves
+          currentState = metadata.BrokerState.FENCED
```
Comment (rondagostino) [inline core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:282]: The heartbeat needs to send a fence-local-broker event to the broker's metadata listener via something like this:

```
kafkaBroker.brokerMetadataListener.put(OutOfBandFenceLocalBrokerEvent(brokerEpoch))
```

**Review by rondagostino — COMMENTED** [review]
Indicated the two spots where the heartbeat needs to notify the broker's metadata listener.

File: core/src/main/scala/kafka/server/BrokerLifecycleManager.scala (hunk 1 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:160]: Done

**Comment by cmccabe** [comment]
LGTM.  Thanks, @aloknnikhil .  I had to rebase this since there were a bunch of merges causing issues with thistory.  So I merged from the command line.

**Pull request closed without merging**